### ISSUE TITLE
How can I set the stop loss at the lowest Low of the previous 5 candles?

### ISSUE THREAD

**Issue by stspinu** [freqtrade/freqtrade#8134]
Hi folks,

Do you know how I can set my stop-loss at the lowest price among the last 5 candles before the entry moment?

It should likely be achieved with **use_custom_stoploss** or **custom_exit()**, but not sure how.

Here is the strategy in 3 simple steps (also illustrated in the screenshot attached)
1. Wait for the candle to cross from above and close below the 20 EMA (green line).
2. Wait for one of the next candles to close above that same 20 EMA line.
3. Enter on the next candle. 

The idea is to have the stop-loss at the lowest price reached between the two crossing candles (from step 1 and step 2) and ideally set a 1:2 risk reward ration (ex: if stop-loss is minus 1%, then make the take profit as positive 2%). 

I am also attaching my strategy file at the bottom (as a code snippet), which shows the entry (let's focus on LONG only).

There are other aspects to the strategy, but I just simplified it here, hoping that I can get some help. If you need more details, let me know and I would love to share more. 

![stopLossFreqTrade](https://user-images.githubusercontent.com/46924453/217285828-2a00e64b-6ebe-44be-bf60-0dd92026cc1d.png)

Thank you very much for your time and help here!

P.S.
This is the full strategy, free on YouTube, for anyone interested:
https://www.youtube.com/watch?v=xkLo5L212ds

```python
# pragma pylint: disable=missing-docstring, invalid-name, pointless-string-statement
# flake8: noqa: F401

# --- Do not remove these libs ---
import numpy as np  # noqa
import pandas as pd  # noqa
from pandas import DataFrame

from freqtrade.strategy import (BooleanParameter, CategoricalParameter, DecimalParameter,
                                IStrategy, IntParameter)

# --------------------------------
# Add your lib to import here
import talib.abstract as ta
import pandas_ta as pta
import freqtrade.vendor.qtpylib.indicators as qtpylib


class ema2050200min15(IStrategy):
    """
    This is a strategy template to get you started.
    More information in https://www.freqtrade.io/en/latest/strategy-customization/

    You can:
        :return: a Dataframe with all mandatory indicators for the strategies
    - Rename the class name (Do not forget to update class_name)
    - Add any methods you want to build your strategy
    - Add any lib you need to build your strategy

    You must keep:
    - the lib in the section "Do not remove these libs"
    - the methods: populate_indicators, populate_entry_trend, populate_exit_trend
    You should keep:
    - timeframe, minimal_roi, stoploss, trailing_*
    """
    # Strategy interface version - allow new iterations of the strategy interface.
    # Check the documentation or the Sample strategy to get the latest version.
    INTERFACE_VERSION = 3

    # Optimal timeframe for the strategy.
    timeframe = '15m'

    # Can this strategy go short?
    can_short: bool = True

    # Minimal ROI designed for the strategy.
    # This attribute will be overridden if the config file contains "minimal_roi".
    minimal_roi = {
        "0": 0.005
    }

    # Optimal stoploss designed for the strategy.
    # This attribute will be overridden if the config file contains "stoploss".
    stoploss = -0.005

    # Trailing stoploss
    trailing_stop = False
    # trailing_only_offset_is_reached = False
    # trailing_stop_positive = 0.01
    # trailing_stop_positive_offset = 0.0  # Disabled / not configured

    # Run "populate_indicators()" only for new candle.
    process_only_new_candles = False

    # These values can be overridden in the config.
    use_exit_signal = True
    exit_profit_only = False
    ignore_roi_if_entry_signal = False

    # Number of candles the strategy requires before producing valid signals
    startup_candle_count: int = 960

    # Strategy parameters
    buy_rsi = IntParameter(10, 40, default=30, space="buy")
    sell_rsi = IntParameter(60, 90, default=70, space="sell")

    # Optional order type mapping.
    order_types = {
        'entry': 'limit',
        'exit': 'limit',
        'stoploss': 'market',
        'stoploss_on_exchange': False
    }

    # Optional order time in force.
    order_time_in_force = {
        'entry': 'gtc',
        'exit': 'gtc'
    }
    
    @property
    def plot_config(self):
        return {
            # Main plot indicators (Moving averages, ...)
            'main_plot': {
                'tema': {},
                'sar': {'color': 'white'},
            },
            'subplots': {
                # Subplots - each dict defines one additional plot
                "MACD": {
                    'macd': {'color': 'blue'},
                    'macdsignal': {'color': 'orange'},
                },
                "RSI": {
                    'rsi': {'color': 'red'},
                }
            }
        }

    def informative_pairs(self):
        """
        Define additional, informative pair/interval combinations to be cached from the exchange.
        These pair/interval combinations are non-tradeable, unless they are part
        of the whitelist as well.
        For more information, please consult the documentation
        :return: List of tuples in the format (pair, interval)
            Sample: return [("ETH/USDT", "5m"),
                            ("BTC/USDT", "15m"),
                            ]
        """
        return []

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Adds several different TA indicators to the given DataFrame

        Performance Note: For the best performance be frugal on the number of indicators
        you are using. Let uncomment only the indicator you are using in your strategies
        or your hyperopt configuration, otherwise you will waste your memory and CPU usage.
        :param dataframe: Dataframe with data from the exchange
        :param metadata: Additional information, like the currently traded pair
        :return: a Dataframe with all mandatory indicators for the strategies
        """
        
        
        # # EMA - Exponential Moving Average
        # dataframe['ema3'] = ta.EMA(dataframe, timeperiod=3)
        # dataframe['ema5'] = ta.EMA(dataframe, timeperiod=5)
        # dataframe['ema10'] = ta.EMA(dataframe, timeperiod=10)
        dataframe['ema20'] = ta.EMA(dataframe, timeperiod=20)
        dataframe['ema50'] = ta.EMA(dataframe, timeperiod=50)
        dataframe['ema200'] = ta.EMA(dataframe, timeperiod=200)

        
        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Based on TA indicators, populates the entry signal for the given dataframe
        :param dataframe: DataFrame
        :param metadata: Additional information, like the currently traded pair
        :return: DataFrame with entry columns populated
        """
        dataframe.loc[
            (
                # UPTREND
                (dataframe['ema20'] > dataframe['ema50']) &
                (dataframe['ema50'] > dataframe['ema200']) &

                # Candle closes above EMA20
                (dataframe['close'] > dataframe['ema20'] ) &

                # One of the previous 5 candles crossed below EMA20
                (
                    ((dataframe['close'].shift(1) < dataframe['ema20'].shift(1)) & (dataframe['open'].shift(1) > dataframe['ema20'].shift(1))) |
                    ((dataframe['close'].shift(2) < dataframe['ema20'].shift(2)) & (dataframe['open'].shift(2) > dataframe['ema20'].shift(2))) |
                    ((dataframe['close'].shift(3) < dataframe['ema20'].shift(3)) & (dataframe['open'].shift(3) > dataframe['ema20'].shift(3))) |
                    ((dataframe['close'].shift(4) < dataframe['ema20'].shift(4)) & (dataframe['open'].shift(4) > dataframe['ema20'].shift(4))) |
                    ((dataframe['close'].shift(5) < dataframe['ema20'].shift(5)) & (dataframe['open'].shift(5) > dataframe['ema20'].shift(5)))
                    
                )
                 
            ),
            'enter_long'] = 1
        # Uncomment to use shorts (Only used in futures/margin mode. Check the documentation for more info)
        
        dataframe.loc[
            (
                # DOWNTREND
                (dataframe['ema20'] < dataframe['ema50']) &
                (dataframe['ema50'] < dataframe['ema200']) &

                # Candle closes below EMA20
                (dataframe['close'] < dataframe['ema20'] ) &

                # One of the previous 5 candles crossed above EMA20
                (
                    ((dataframe['close'].shift(1) > dataframe['ema20'].shift(1)) & (dataframe['open'].shift(1) < dataframe['ema20'].shift(1))) |
                    ((dataframe['close'].shift(2) > dataframe['ema20'].shift(2)) & (dataframe['open'].shift(2) < dataframe['ema20'].shift(2))) |
                    ((dataframe['close'].shift(3) > dataframe['ema20'].shift(3)) & (dataframe['open'].shift(3) < dataframe['ema20'].shift(3))) |
                    ((dataframe['close'].shift(4) > dataframe['ema20'].shift(4)) & (dataframe['open'].shift(4) < dataframe['ema20'].shift(4))) |
                    ((dataframe['close'].shift(5) > dataframe['ema20'].shift(5)) & (dataframe['open'].shift(5) < dataframe['ema20'].shift(5)))
                    
                )
            ),
            'enter_short'] = 1
        

        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Based on TA indicators, populates the exit signal for the given dataframe
        :param dataframe: DataFrame
        :param metadata: Additional information, like the currently traded pair
        :return: DataFrame with exit columns populated
        """
        dataframe.loc[
            (
                (dataframe['ema20'] < -1)  # Make sure Volume is not 0
            ),
            'exit_long'] = 1
        # Uncomment to use shorts (Only used in futures/margin mode. Check the documentation for more info)
        """
        dataframe.loc[
            (
                (qtpylib.crossed_above(dataframe['rsi'], self.buy_rsi.value)) &  # Signal: RSI crosses above buy_rsi
                (dataframe['tema'] <= dataframe['bb_middleband']) &  # Guard: tema below BB middle
                (dataframe['tema'] > dataframe['tema'].shift(1)) &  # Guard: tema is raising
                (dataframe['volume'] > 0)  # Make sure Volume is not 0
            ),
            'exit_short'] = 1
        """
        return dataframe
    ```

**Comment by xmatthias** [comment]
well i guess you can use something like `dataframe.rolling(5).min()` within the `custom_stoploss()` callback.

Obviously the "5" will have to be dynamic based on the actual entry signal candle - and you'll need to ensure you're not simply using the "min last 5" all the time - as `custom_stoploss()` will be called throughout the trade's life.

**Comment by stspinu** [comment]
> well i guess you can use something like `dataframe.rolling(5).min()` within the `custom_stoploss()` callback.
> 
> Obviously the "5" will have to be dynamic based on the actual entry signal candle - and you'll need to ensure you're not simply using the "min last 5" all the time - as `custom_stoploss()` will be called throughout the trade's life.

Thank you Matthias for the very quick reply. Indeed, I need to make it a bit dynamic.

Just as you said -> "**_I'll need to ensure I'm not simply using the "min last 5" all the time_**" 

In order to control what stop_loss value is returned whenever the custom_stoploss() method is called, I would need to know what is the stoploss ratio at any specific moment. 

Could you please tell me what is the **stoploss** variable (the one that gets the value returned by _custom_stoploss()_ function)?

Then I can just define some logic to handle it.

Unfortunately, it's not the **self.stoploss** variable (that one is always hard-coded from the beginning), so it must be some other variable and I would appreciate if you could tell me which one it is. Thank you, again!

**Comment by xmatthias** [comment]
well the trade object has both a `trade.stop_loss` attribute (absolute stoploss) - as well as the corresponding relative stoploss (`trade.stop_loss_pct`) - which correspond to the stoploss that's currently active for that trade.

**Comment by stspinu** [comment]
Perfect, this is pretty much all I needed - I should be able to take it from here. Thank you very much!

**Comment by stspinu** [comment]
Closing this, thanks!

**Comment by arun-tandon** [comment]
@stspinu 
Can you share the updated code with stop loss at the lowest Low of the previous 5 candles?

**Comment by stspinu** [comment]
> @stspinu Can you share the updated code with stop loss at the lowest Low of the previous 5 candles?

hi Arun, I tried to look for it, but couldn't find it.

However, I think you should tell the custom_stoploss() function to choose the smallest (minimum) value of the 'low' column from the last 5 rows of the dataframe object.

It's most likely something like __dataframe['low'].iloc[-5:].min()__

__dataframe['low'].iloc[-5:].min()__ should take the dataframe object, look from the 5th last element onward and select the minimum value. Not sure if this is the syntax, but you can use Google or Chat GPT to look it up. 

However, I think the custom_stoploss() function needs to return something like trade.stop_loss_pct... not sure where that 'pct' (probably percentage) is coming from or why I had to use it. This happened a while ago, so I have vague recollections, but I remember it gave me some headaches before I figured it out.  Let me know if I can help further.

**Comment by elja** [comment]
@stspinu can you access dataframe within custom_stoploss? I don't think that it receives it as an argument.

**Comment by xmatthias** [comment]
@elja you can through [get_analyzed_dataframe](https://www.freqtrade.io/en/latest/strategy-customization/#get_pair_dataframepair-timeframe) ... (make sure to handle not getting a response - which MAY happen right after a restart).
